### PR TITLE
Add unversioned.MicroTime type

### PR DESCRIPTION
**Pull request by gmarek**
Ref. #36752

cc @smarterclayton @wojtek-t @dchen1107 @timothysc @bgrant0607

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37212)

### CODE REVIEW

**Comment by dims** [comment]
@gmarek : why RFC3339 (instead of ISO 8601?)

**Comment by gmarek** [comment]
Because XYZ - in other words, I have no idea, and I don't care, it's just the thing we have now:)

**Comment by dims** [comment]
LOL

**Comment by smarterclayton** [comment]
RFC3339 is much more commonly used in the REST space than ISO8601, and more libraries and clients (unscientific study) use it or expect it.

**Comment by smarterclayton** [comment]
Changes are ok to me - we can dry them up later.

**Comment by dims** [comment]
Thanks @smarterclayton

File: pkg/api/unversioned/generated.proto [hunk 1]
```diff
@@ -232,6 +232,24 @@ message ListMeta {
   optional string resourceVersion = 2;
 }
 
+// MicroTime is version of Time with microsecond level precision.
+// 
+// +protobuf.options.marshal=false
+// +protobuf.as=Timestamp
+// +protobuf.options.(gogoproto.goproto_stringer)=false
+message MicroTime {
```
Comment (timothysc) [inline pkg/api/unversioned/generated.proto]: Isn't this ~= https://github.com/google/protobuf/blob/master/src/google/protobuf/timestamp.proto 

Just wondering why we're duping.

Comment (gmarek) [inline pkg/api/unversioned/generated.proto]: For consistency with the rest of the system. I didn't want to spend time to check what I'd need to do to serialize our custom go/json to this proto.

**Comment by smarterclayton** [comment]
We can't change the proto serialization anyway (now that it's locked).

**Comment by smarterclayton** [comment]
Also, there are bugs in go-proto with google types and they were only
recently fixed.

**Comment by gmarek** [comment]
@lavalamp @smarterclayton - can we move forward with this?

**Comment by gmarek** [comment]
@smarterclayton - done.

**Comment by gmarek** [comment]
@smarterclayton @lavalamp PTAL

**Review by smarterclayton — APPROVED** [review]

**Comment by smarterclayton** [comment]
Rebase otherwise lgtm

**Comment by gmarek** [comment]
GKE tests are broken now.

File: pkg/api/unversioned/time.go [hunk 2]
```diff
@@ -79,6 +80,16 @@ func (t Time) Equal(u Time) bool {
 	return t.Time.Equal(u.Time)
 }
 
+// BeforeMicro reports whether the time instant t is before micro instance u.
```
Comment (lavalamp) [inline pkg/api/unversioned/time.go]: What is a "micro instance"?

File: pkg/api/unversioned/time.go [hunk 3]
```diff
@@ -79,6 +80,16 @@ func (t Time) Equal(u Time) bool {
...
+func (t Time) BeforeMicro(u v1.MicroTime) bool {
+	return t.Time.Before(u.Time)
```
Comment (lavalamp) [inline pkg/api/unversioned/time.go]: Shouldn't this round (er, truncate?) to nearest microsecond before comparing? Same for below function. If this function doesn't do anything special then maybe just delete it?

File: pkg/apis/meta/v1/micro_time.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	return t.Time.Equal(u.Time)
+}
+
+// UnixMicro returns the local time corresponding to the given Unix time
+// by wrapping time.Unix.
+func UnixMicro(sec int64, nsec int64) MicroTime {
+	return MicroTime{time.Unix(sec, nsec)}
+}
+
+// Rfc3339Copy returns a copy of the MicroTime at second-level precision.
```
Comment (lavalamp) [inline pkg/apis/meta/v1/micro_time.go]: second-level?

File: pkg/apis/meta/v1/micro_time.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+// EqualTime reports whether the time instant t is equal to second-lever precision u.
+func (t MicroTime) EqualTime(u Time) bool {
+	return t.Time.Equal(u.Time)
+}
+
+// UnixMicro returns the local time corresponding to the given Unix time
+// by wrapping time.Unix.
+func UnixMicro(sec int64, nsec int64) MicroTime {
+	return MicroTime{time.Unix(sec, nsec)}
```
Comment (lavalamp) [inline pkg/apis/meta/v1/micro_time.go]: shouldn't you truncate to microseconds?

File: pkg/apis/meta/v1/micro_time.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+// DateMicro returns the MicroTime corresponding to the supplied parameters
+// by wrapping time.Date.
+func DateMicro(year int, month time.Month, day, hour, min, sec, nsec int, loc *time.Location) MicroTime {
+	return MicroTime{time.Date(year, month, day, hour, min, sec, nsec, loc)}
+}
+
+// NowMicro returns the current local time.
+func NowMicro() MicroTime {
+	return MicroTime{time.Now()}
```
Comment (lavalamp) [inline pkg/apis/meta/v1/micro_time.go]: shouldn't you truncate to microseconds?

File: pkg/apis/meta/v1/micro_time.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+// NewMicroTime returns a wrapped instance of the provided time
+func NewMicroTime(time time.Time) MicroTime {
+	return MicroTime{time}
+}
+
+// DateMicro returns the MicroTime corresponding to the supplied parameters
+// by wrapping time.Date.
+func DateMicro(year int, month time.Month, day, hour, min, sec, nsec int, loc *time.Location) MicroTime {
+	return MicroTime{time.Date(year, month, day, hour, min, sec, nsec, loc)}
```
Comment (lavalamp) [inline pkg/apis/meta/v1/micro_time.go]: shouldn't you truncate to microseconds?

File: pkg/apis/meta/v1/micro_time.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+// String returns the representation of the time.
+func (t MicroTime) String() string {
+	return t.Time.String()
+}
+
+// NewMicroTime returns a wrapped instance of the provided time
+func NewMicroTime(time time.Time) MicroTime {
+	return MicroTime{time}
```
Comment (lavalamp) [inline pkg/apis/meta/v1/micro_time.go]: shouldn't you truncate to microseconds?

File: pkg/apis/meta/v1/micro_time.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+	return t.UTC().Format(RFC3339Micro), nil
+}
+
+// Fuzz satisfies fuzz.Interface.
+func (t *MicroTime) Fuzz(c fuzz.Continue) {
+	if t == nil {
+		return
+	}
+	// Allow for about 1000 years of randomness.  Leave off nanoseconds
```
Comment (lavalamp) [inline pkg/apis/meta/v1/micro_time.go]: Update this. It should randomize to microseconds.

File: pkg/apis/meta/v1/generated.proto [hunk 10]
```diff
@@ -245,6 +245,24 @@ message ListMeta {
...
+  // Represents seconds of UTC time since Unix epoch
+  // 1970-01-01T00:00:00Z. Must be from from 0001-01-01T00:00:00Z to
+  // 9999-12-31T23:59:59Z inclusive.
+  optional int64 seconds = 1;
+
+  // Non-negative fractions of a second at nanosecond resolution. Negative
+  // second values with fractions must still have non-negative nanos values
+  // that count forward in time. Must be from 0 to 999,999,999
+  // inclusive. This field may be limited in precision depending on context.
+  optional int32 nanos = 2;
```
Comment (lavalamp) [inline pkg/apis/meta/v1/generated.proto]: Why do you store nanoseconds for a microtime type? That doesn't make much sense.

File: pkg/apis/meta/v1/time.go [hunk 11]
```diff
@@ -83,6 +83,16 @@ func (t Time) Equal(u Time) bool {
 	return t.Time.Equal(u.Time)
 }
 
+// BeforeMicro reports whether the time instant t is before micro instance u.
```
Comment (lavalamp) [inline pkg/apis/meta/v1/time.go]: Same comment as other file, delete these or make them round correctly to micros.

**Review by lavalamp — CHANGES_REQUESTED** [review]
Why do you need this type at all? If it is really necessary, then please make it behave the way the name advertises (or change the name).

**Comment by lavalamp** [comment]
And very sorry for delay in review. I was OOO for a long time last month.

File: pkg/api/unversioned/time.go (hunk 2 above)
Comment (gmarek) [inline pkg/api/unversioned/time.go]: Rephrased.

File: pkg/api/unversioned/time.go (hunk 3 above)
Comment (gmarek) [inline pkg/api/unversioned/time.go]: This shouldn't be needed - both Time and MicroTime, as go-structs have full precision. The only difference is that when serialized to JSON the latter is more precise.

File: pkg/apis/meta/v1/micro_time.go (hunk 8 above)
Comment (gmarek) [inline pkg/apis/meta/v1/micro_time.go]: We're not doing it in normal 'Time' which has a second precision. I don't see why we should do it here.

File: pkg/apis/meta/v1/micro_time.go (hunk 7 above)
Comment (gmarek) [inline pkg/apis/meta/v1/micro_time.go]: We're not doing it in normal 'Time' which has a second precision. I don't see why we should do it here.

File: pkg/apis/meta/v1/micro_time.go (hunk 6 above)
Comment (gmarek) [inline pkg/apis/meta/v1/micro_time.go]: We're not doing it in normal 'Time' which has a second precision. I don't see why we should do it here.

File: pkg/apis/meta/v1/micro_time.go (hunk 5 above)
Comment (gmarek) [inline pkg/apis/meta/v1/micro_time.go]: We're not doing it in normal 'Time' which has a second precision. I don't see why we should do it here.

File: pkg/apis/meta/v1/micro_time.go (hunk 4 above)
Comment (gmarek) [inline pkg/apis/meta/v1/micro_time.go]: Sorry, I missed that.

File: pkg/apis/meta/v1/micro_time.go (hunk 9 above)
Comment (gmarek) [inline pkg/apis/meta/v1/micro_time.go]: Thx.

File: pkg/apis/meta/v1/time.go (hunk 11 above)
Comment (gmarek) [inline pkg/apis/meta/v1/time.go]: I don't think we should do that. Unversioned Time is used internally as normal timestamp with nanosecond precision, the only thing that I'm not happy with is that when serialized it truncates time to second. The second 'MicroTime' is designed to be exactly the same, but serialized differently. Hence actually we should compare 'full' representations.

File: pkg/apis/meta/v1/generated.proto (hunk 10 above)
Comment (gmarek) [inline pkg/apis/meta/v1/generated.proto]: Again - this is exactly the same thing as we have for Time.

**Comment by gmarek** [comment]
Applied few comments and replied to the rest - @lavalamp PTAL.

**Comment by lavalamp** [comment]
I still don't understand-- why do you need this type at all?

File: pkg/apis/meta/v1/micro_time.go (hunk 8 above)
Comment (lavalamp) [inline pkg/apis/meta/v1/micro_time.go]: If you don't want to round to micros, then I really don't understand why you want a separate type at all.

File: pkg/apis/meta/v1/time.go (hunk 11 above)
Comment (lavalamp) [inline pkg/apis/meta/v1/time.go]: What bad thing happens if you update the normal time to store micros instead of seconds?

Comment (smarterclayton) [inline pkg/apis/meta/v1/time.go]: We break API compatibility with v1.

Comment (gmarek) [inline pkg/apis/meta/v1/time.go]: @lavalamp - as @smarterclayton we can't do that, however tempting it would be.

**Comment by lavalamp** [comment]
I still don't understand why we need this type at all?

**Comment by lavalamp** [comment]
I feel really bad for sitting on this PR for so long. I guess primarily two things:
a) Why do we need this type? I guess you probably have a good reason to store micros? But is there any way to e.g. parameterize the current type instead of copying it?
b) I still don't like the fact that this has got "micro" in the name but most of the methods actually don't constrain anything to micros. As in, it will behave drastically differently depending on whether you use times that you generated locally vs after they've been through a serialization/deserialization loop.

I almost feel bad enough (about the delay in me looking at this) to approve it anyway, but not quite :)

**Comment by gmarek** [comment]
The reason why we need micros is that second granularity is by far not enough to allow... well pretty much anything on big clusters (you get unsorted blob of hundreds of things that happened in a given second). For me personally it's problematic because I need to use those timestamps to trace e2e pod startup latency.

The problem you mention (not being micro everywhere) already exists with default time, which has exactly the same property, but with second granularity. I don't think it ever caused problems, so I don't really think that solving it in some elaborate way makes a lot of sense.

**Comment by gmarek** [comment]
@lavalamp

**Comment by lavalamp** [comment]
Last question, still would like an answer to "but is there any way to e.g. parameterize the current type instead of copying it?"

I'm going to remove myself from this PR and stop blocking it, I have too many other priorities.

**Review by lavalamp — APPROVED** [review]

**Comment by ethernetdan** [comment]
Removing PR from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

**Comment by timothysc** [comment]
@gmarek I'd like to restart this PR if possible, given our inability to trace across components easily. 

/cc @kensimon

**Comment by gmarek** [comment]
Yes, I agree - I was planning to restart this, but I was busy with reviews and playing with OT (I can trace API requests from client to the API server and back, so I'll probably start sending PRs soon).

**Comment by gmarek** [comment]
@timothysc @lavalamp @smarterclayton - can we come back to this PR and try to get at least microsecond-timestamps in Events in 1.7?

**Comment by timothysc** [comment]
> can we come back to this PR and try to get at least microsecond-timestamps in Events in 1.7?

+1 from me.

**Comment by gmarek** [comment]
@smarterclayton

**Comment by gmarek** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gmarek** [comment]
@k8s-bot pull-kubernetes-unit test this

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by gmarek** [comment]
Adding do-not-merge label, as @wojtek-t lgtm acts as approve.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time.go [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+// UnixMicro returns the local time corresponding to the given Unix time
+// by wrapping time.Unix.
+func UnixMicro(sec int64, nsec int64) MicroTime {
+	return MicroTime{time.Unix(sec, nsec)}
+}
+
+// Rfc3339Copy returns a copy of the MicroTime at microsecond-level precision.
+func (t MicroTime) Rfc3339MicroCopy() MicroTime {
```
Comment (thockin) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time.go]: How is this different from DeepCopy?

**Review by thockin — COMMENTED** [review]
I don't see obvious problems with this...

@smarterclayton

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time.go (hunk 12 above)
Comment (gmarek) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time.go]: Frankly, I don't remember. Probably copied it from 'Time' at some point in time. Removed.

**Comment by gmarek** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by wojtek-t** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gmarek** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gmarek** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gmarek** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gmarek** [comment]
@k8s-bot pull-kubernetes-unit test this

**Pull request merged**